src/Native/System.Net.Security.Native/pal_ntlmapi.cpp

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "pal_types.h"
+#include "pal_ntlmapi.h"
+#include "pal_utilities.h"
+#include <stddef.h>
+#include <assert.h>
+#include <time.h>
+#include "heimntlm.h"
+#include "openssl/hmac.h"
+#include "openssl/evp.h"
+
+static_assert(PAL_NTLMSSP_NEGOTIATE_UNICODE == NTLM_NEG_UNICODE, "");
+static_assert(PAL_NTLMSSP_REQUEST_TARGET == NTLM_NEG_TARGET, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_SIGN == NTLM_NEG_SIGN, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_SEAL == NTLM_NEG_SEAL, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_NTLM == NTLM_NEG_NTLM, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_ALWAYS_SIGN == NTLM_NEG_ALWAYS_SIGN, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY == NTLM_NEG_NTLM2_SESSION, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_128 == NTLM_ENC_128, "");
+static_assert(PAL_NTLMSSP_NEGOTIATE_KEY_EXCH == NTLM_NEG_KEYEX, "");
+
+const int32_t MD5_DIGEST_LENGTH = 16;
+
+static inline int32_t NetSecurity_SetBufferLength(int32_t status, ntlm_buf** bufferHandle, int32_t* outLength)
+{
+    if (status != 0)
+    {
+        delete *bufferHandle;
+        *bufferHandle = nullptr;
+        *outLength = 0;
+        return status;
+    }
+    *outLength = static_cast<int32_t>((*bufferHandle)->length);
```

Comment:
It looks like ntlm_buf.length is a size_t.  Won't this potentially be truncating on 64-bit?